File: cs_src/DataUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace smake
{
    static class DataUtils
    {
        public static IEnumerable<IEnumerable<string>> CreateCandidates(IEnumerable<string> words, HashSet<string> sws)
        {
            var buffer = new List<string>();
            var candidates = new List<List<string>>();

            foreach (var word in words)
            {
                if (sws.Contains(word.ToLower()))
                {
                    if (buffer.Count > 0)
                    {
                        if (!WordHeuristics(buffer)) 
                        { candidates.Add(buffer); }
                        buffer = new List<string>(); 
                    }
                }
                else if (Regex.Match(word, @"\p{Lu}.+").Success)
                {
                    if (buffer.Count > 0)
                    {
                        if (!WordHeuristics(buffer)) 
                        { candidates.Add(buffer); }
                        buffer = new List<string>();
                    }
                    buffer.Add(word.ToLower());
                }
                else
                {
                    buffer.Add(word.ToLower());
                }
            }
            return candidates;
        }

        private static bool WordHeuristics(List<string> buffer)
        {
            return Regex.Match(buffer[^1], @".*ed").Success || buffer[0].Length <= 3;
        }

        public static (Dictionary<string, int>, Dictionary<string, int>) ComputeFrqDeg(IEnumerable<IEnumerable<string>> candidates)
        {
            var frqMap = new Dictionary<string, int>();
            var degMap = new Dictionary<string, HashSet<string>>();

            foreach (var cset in candidates)
            {
                foreach (var w1 in cset)
                {
                    frqMap.Add1OrIncrement(w1);

                    if (!degMap.ContainsKey(w1))
                    {
                        degMap.Add(w1, new HashSet<string>());
                    }

                    foreach (var w2 in cset) { degMap[w1].Add(w2); }
                }
            }

            var deg = new Dictionary<string, int>();
            foreach (var degPair in degMap)
            {
                deg.Add(degPair.Key, degPair.Value.Count);
            }

            return (frqMap, deg);
        }
    }
}

File: py_src/data_utils.py
import regex as re

def compute_frq_deg(candidates):
    frq_map = dict()
    deg_map = dict() 

    for cset in candidates:
        for w1 in cset:
            if w1 not in frq_map:
                frq_map[w1] = 0
            frq_map[w1] += 1

            if w1 not in deg_map:
                deg_map[w1] = set()

            for w2 in cset:
                deg_map[w1].add(w2)

    deg = {x:len(v) for x,v in deg_map.items() }

    return (frq_map, deg)


r2 = re.compile(r".*ed")
def word_heuristics(candidate):
    return re.match(r2, candidate[-1]) or len(candidate[0]) <= 3

r3 = re.compile(r"\p{Lu}.+")
def create_candidates(words, sws):
    buffer = []
    candidates = []

    for word in words:
        if word.lower() in sws:
            if len(buffer) != 0:
                if not word_heuristics(buffer):
                    candidates.append(buffer)
                buffer = []
        elif re.match(r3, word): 
            if len(buffer) != 0:
                if not word_heuristics(buffer):
                    candidates.append(buffer)
                buffer = []
            buffer.append(word.lower())
        else:
            buffer.append(word.lower())

    return candidates
